src/PrettyPrinting.jl

```diff
@@ -1401,4 +1403,158 @@ function with_unicode(f::Function)
+    ioi = IOCustom(_io)
+    print(ioi, "IOCustom:", Indent())
+    print(ioi, "\nIO: "); show(ioi, io.io)
+    print(ioi, "\nIndent string: \"", io.indent_str, "\"")
+    print(ioi, "\nIndent: ", io.indent_level)
+end
+
+pipe_reader(io::IOCustom) = io.io
+pipe_writer(io::IOCustom) = io.io
+lock(io::IOCustom) = lock(io.io)
+unlock(io::IOCustom) = unlock(io.io)
+
+Base.displaysize(io::IOCustom) = displaysize(io.io)
+
+write(io::IO, ::Indent) = (_unwrap(io).indent_level += 1; 0)
+print(io::IO, ::Indent) = write(io, Indent())
+write(io::IO, ::Dedent) = (_unwrap(io).indent_level = max(0, io.indent_level - 1); 0)
+print(io::IO, ::Dedent) = write(io, Dedent())
+write(io::IO, ::Lowercase) = (_unwrap(io).lowercasefirst = true; 0)
+print(io::IO, ::Lowercase) = write(io, Lowercase())
+write(io::IO, ::Uppercase) = (_unwrap(io).lowercasefirst = false; 0)
+print(io::IO, ::Uppercase) = write(io, Uppercase())
+
+_align_length(io) = length(io.aligns) == 0 ? 0 : io.aligns[end]
+
+write_indent(io::IO) = write(_unwrap(io).io, io.indent_str^io.indent_level)
+
+write(io::IOCustom, chr::Char) = write(io, string(chr)) # Need to catch writing a '\n'
+
+function _write_line(io::IOCustom, str::AbstractString)
+  written = 0
+  c = displaysize(io)[2]
+  ind = io.indent_level * textwidth(io.indent_str)
+  # there might be already something written
+  if c - ind - io.printed < 0
+    spaceleft = mod(c - ind - io.printed, c)
+  else
+    spaceleft = c - ind - io.printed
+  end
+  #@show spaceleft
+  firstlen = min(spaceleft, length(str))
+  firststr = str[1:firstlen]
+  if io.lowercasefirst
+    written += write(io.io, lowercasefirst(firststr))
+  else
+    written += write(io.io, firststr)
+    io.lowercasefirst = false
+  end
+  io.printed += textwidth(firststr)
+  reststr = str[firstlen + 1:end]
+  it = Iterators.partition(1:textwidth(reststr), c - ind > 0 ? c - ind : mod(c - ind, c))
+  for i in it
+    written += write(io.io, "\n")
+    written += write_indent(io)
+    written += write(io.io, reststr[i])
+    io.printed = textwidth(reststr[i])
+  end
+  return written
+end
+
+function write(io::IOCustom, str::String)
+  written = 0
+  if str == "\n"
+    written = write(io.io, str)
+    io.indented_line = false
+    return written
+  end
+
+  for (i, line) in enumerate(split(str, "\n"))
+    if i != 1
+      written += write(io.io, "\n")
+      io.indented_line = false
+    end
+
+    if !io.indented_line
+      written += write_indent(io)
+      io.indented_line = true
+    end
+    written += _write_line(io, line)
+  end
+  if !isempty(str) && last(str) == '\n'
+    io.indented_line = false
+  else
+    io.indented_line = true
+  end
+  return written
+end
+
+Base.write(io::IOContext{<: IOCustom}, s::Union{SubString{String}, String}) = Base.write(Base.unwrapcontext(io)[1], s)
+
+# println(io) redirects to print(io, '\n')
+Base.write(io::IOContext{<: IOCustom}, s::Char) = Base.write(Base.unwrapcontext(io)[1], s)
+
+
+pretty(io::IO) = IOCustom(io)
```

Comment:
Perhaps make this idempotent for `IOCustom` inputs?
```suggestion
pretty(io::IO) = IOCustom(io)
pretty(io::IOCustom) = io
```

Reply:
It is already because of `IOCustom(io::IOCustom) = io` a few lines above.